RFE: debugging ability

Palatis opened the pull request:
1. currently it only checks "parser not advancing" in `ZeroOrMoreRule`, `OneOrMoreRule`, and `CountedRule` on a DEBUG build.
however, nuget packages are always build with RELEASE configuration, guarding these checks with `#if DEBUG` make it harder to diagnostic what's going on under the hood - the client code just hit an infinite loop.
thus, just throw the `ParserException` for RELEASE builds.

2. enable the ability to trace rule matching process in an RELEASE build
similar reason to the previous one, nuget packages are build with RELEASE configuration...

well, the assumption that "nuget packages are always RELEASE builds" might not be that correct.
they can be build in DEBUG, however packages on nuget.org are RELEASE, anyway.

Palatis committed:
output rule match trace logs to `Debug` instead of `Console`

and only do that when Debugger is attached.

cdiggins commented on Parakeet/Rule.cs:
```diff
@@ -491,15 +484,10 @@ protected override ParserState MatchImplementation(ParserState state)
-                    {
-                        throw new ParserException(curr, "Parser is no longer making progress");
-                    }
+                if (Debugger.IsAttached && next != null && next.Position <= curr.Position)
```
The `Debugger.IsAttached` has no real benefit here, if we remove the conditional compilation might as well do the full check and throw an exception. A performance hit will occur.

Palatis replied:
i think you're right.

however during normal operation this will just be a null-ness check.

cdiggins commented on Parakeet/Rule.cs:
```diff
@@ -25,17 +25,15 @@ public abstract class Rule
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public ParserState Match(ParserState state)
         {
-#if DEBUG
```
This will have a small performance impact. I spent a lot of time trying to make the parser as fast as possible. That said, at this point I think the parser is plenty fast for its primary use case.

Palatis replied:
i think it may be worth creating the debug version of these classes, for example `DebugZeroOrMoreRule`, `DebugOneOrMoreRule`, and `DebugCountedRule`, then select which version of the rule instance to create during grammar construction process.
thus we can minimize the performance impact during RELEASE builds.

```CSharp
public class DebugOneOrMoreRule : OneOrMoreRule
{
    public override ParserState MatchImplementation(ParserState state)
    {
        /* same implementation, but check for not advancing */
    }
}

public abstract class Grammar : IGrammar
{
    public virtual bool IsDebugging => false;
    public Rule OneOrMore(Rule r)
        => IsDebugging ? new DebugOneOrMoreRule(r) : new OneOrMoreRule(r);
}

public static class RuleExtensions
{
    public static Rule OneOrMore(Rule rule, bool debugging = false) ?
        => debugging ? new DebugOneOrMoreRule(rule) : new OneOrMoreRule(rule);
}
```

```CSharp
public sealed class RealGrammar : Grammar
{
#if DEBUG
    public override bool IsDebugging => true;
#endif

    public Rule S => Node(...);
    public Rule T => Named(OneOrMore(S));
    public Rule R => Named(T.OneOrMore(IsDebugging));
}
```

something like that.

thus we move the performance impact from parsing to rule-creation.

cdiggins reviewed (changes requested):
I think that the improvements in usability merit the performance hit. Just remove the "if debugger attached" in the spot where I recommend.

Palatis commented on Parakeet/Rule.cs:
```diff
@@ -293,23 +291,23 @@ public static bool[] CharsToTable(IEnumerable<char> chars)
-        public CharSetRule(params char[] chars) 
-            : this(CharsToTable(chars)) 
+        public CharSetRule(params char[] chars)
+            : this(CharsToTable(chars))
```
weird, i don't see those line-ending whitespace diffs on my local tree...

wait, i'll fix that.

Palatis committed:
report ParserException when not making progress in RELEASE build

nuget packages are always build with RELEASE configuration. guard these checks with `#if DEBUG` makes it harder to diagnostic what's going on under the hood - the client code just hit an infinite loop.

Palatis commented:
yeah this looks more like it.
hate mixed line-endings...

cdiggins reviewed (approved):
Good job, thanks.
